lib/sigstore/verifier.rb

```diff
@@ -61,6 +57,27 @@ def verify(materials:, policy:)
       # The Verifier MUST then parse the integratedTime as a Unix timestamp (seconds since January 1, 1970 UTC).
       # If verification or timestamp parsing fails, the Verifier MUST abort.
 
+      begin
+        # TODO: should this instead be an input to the verify method?
+        # See https://docs.google.com/document/d/1kbhK2qyPPk8SLavHzYSDM8-Ueul9_oxIMVFuWMWKz0E/edit?disco=AAABQVV-gT0
+        entry = materials.find_rekor_entry(@rekor_client)
+      rescue Sigstore::Error::MissingRekorEntry
+        return VerificationFailure.new("Rekor entry not found")
+      else
+        if entry.inclusion_proof&.checkpoint
+          Internal::Merkle.verify_merkle_inclusion(entry)
+          Rekor::Checkpoint.verify_checkpoint(@rekor_client, entry)
+        elsif !materials.offline
+          return VerificationFailure.new("Missing Rekor inclusion proof")
+        else
+          warn "inclusion proof not present in bundle: skipping due to offline verification"
+        end
+      end
+
+      Internal::SET.verify_set(client: @rekor_client, entry: entry) if entry.inclusion_promise
+
+      timestamps << Time.at(entry.integrated_time).utc
```

Comment:
Collecting timestamps from multiple sources looks fine to me, but it results in the funkiness below with multiple chain candidates. If you wanted to reduce the search space here, another option would possibly be to check all timestamps for consistency, i.e. fail if any signed source of time exceeds the validity window of the leaf certificate.

(This isn't documented anywhere in the client spec though, so maybe it requires determination there first.)

Reply:
I don't think that's sufficient, since you could fail to verify a chain at a timestamp where the leaf is valid?

Reply:
iirc the spec specifically calls out that you do x509 validation multiple times